### ISSUE TITLE
Support 2023.x Awaitable

### ISSUE THREAD

**Issue by bdovaz** [microsoft/Microsoft.Unity.Analyzers#255]
#### Problem statement

Since alpha 2023.1, Unity has added a new type, Awaitable. With this type you can do operations that were previously done with coroutines such as: wait N seconds, wait a frame, switch to background / main thread...

It is supported in the messages known by Unity from a MonoBehaviour such as Awake, Start, Update...

_**I know it's early but the sooner we anticipate the better.**_

Context:

https://docs.unity3d.com/2023.1/Documentation/ScriptReference/Awaitable.html

https://docs.unity3d.com/2023.1/Documentation/ScriptReference/AwaitableCompletionSource.html

Now reports this false positive:

![image](https://user-images.githubusercontent.com/950602/208371999-e41fa9c8-6a48-4657-a87d-37fff550c8e1.png)

#### Proposed solution

That this code is valid for the analyzer:

```csharp
public class NewBehaviourScript : MonoBehaviour
{
    async Awaitable Awake()
    {
        Debug.Log("1");
        await Awaitable.WaitForSecondsAsync(1);
        Debug.Log("2");
    }
}
```

**Edit:** Seeing this answer I see that it is supported but it is not recommended, you would have to discuss it with Unity just in case: https://forum.unity.com/threads/unity-future-net-development-status.1092205/page-30#post-8674929

**Comment by sailro** [comment]
Thank you for your feedback !

- `UNT0006` should support `Awaitable` given in the end the method will be called by the Unity runtime (that was the goal of this rule to be sure callbacks are effectively called). And now this diagnostic severity is `Warning`.

- We can add another diagnostic to check that Unity messages invoked by the runtime are not leaking (so using async void and not async Awaitable ?), as an `Info` severity.

What do you think @simonferquel ?

**Comment by simonferquel** [comment]
2 notes here:
- Unfortunately, VoidAsyncMethodBuilder seem to allocate unfortunately - at least in Mono, so I'd prefer we avoid async void methods
- Returning an Awaitable that is not awaited, will prevent it to return to the Awaitable pool, leading to undesired allocations as well (they seem a little less dramatic, but still not desirable)

A 3rd option, is to use UniTaskVoid (from UniTask package) which provides an elegant way to avoid allocating in that case.

If popular requests arise, we'll probably work on adding a similar type as UniTaskVoid.

Anyway, something to keep in mind, is that no matter what allocations are there or not, async / await is not free - it introduces code refactoring that can remove inlining possibilities etc. We should then promote patterns where
- we avoid lots of concurrent async methods running at the same time
- we avoid over-composition of async methods (async methods calling a deep graph of async methods)

Thus some simple rules like, warning on Awaitable return type on "Update" message, while accepting it on "Start" or "Awake" would make a lot of sense

**Comment by christallire** [comment]
Unity 6 is now in production. Let's do this.

**Comment by sailro** [comment]
> Unity 6 is now in production. Let's do this.

Indeed. I'm crafting a PR.

**Comment by christallire** [comment]
Would it be very helpful to support "UniTask" or any variant of Awaitable.

**Comment by sailro** [comment]
Yes, let's iterate on this. First we'll see how it goes with the built-in support.
